network/p2p/scoring/subscription_validator_test.go

```diff
@@ -0,0 +1,314 @@
+	sp.On("GetSubscribedTopics", peer1).Return([]string{})
+
+	// as peer 1 has not subscribed to any topic, the subscription validator should return no error.
+	require.NoError(t, sv.CheckSubscribedToAllowedTopics(peer1))
+}
+
+// TestSubscriptionValidator_UnknownIdentity tests that when a peer does not have an authorized identity,
+// the subscription validator returns an error.
+func TestSubscriptionValidator_UnknownIdentity(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	peer := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer).Return(nil, false)
+	sp.On("GetSubscribedTopics", peer).Return([]string{})
+
+	err := sv.CheckSubscribedToAllowedTopics(peer)
+	require.Error(t, err)
+	require.True(t, scoring.IsInvalidPeerIDError(err))
+	require.Contains(t, err.Error(), scoring.PeerIdStatusUnknown)
+}
+
+// TestSubscriptionValidator_EjectedIdentity tests that when a peer has been ejected, the subscription
+// validator returns an error.
+func TestSubscriptionValidator_EjectedIdentity(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	ejectedIdentity := unittest.IdentityFixture()
+	ejectedIdentity.Ejected = true
+	peer := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer).Return(ejectedIdentity, true)
+	sp.On("GetSubscribedTopics", peer).Return([]string{})
+
+	err := sv.CheckSubscribedToAllowedTopics(peer)
+	require.Error(t, err)
+	require.True(t, scoring.IsInvalidPeerIDError(err))
+	require.Contains(t, err.Error(), scoring.PeerIdStatusEjected)
+}
+
+// TestSubscriptionValidator_UnknownChannel tests that when a peer has subscribed to an unknown
+// topic, the subscription validator returns an error.
+func TestSubscriptionValidator_UnknownChannel(t *testing.T) {
+	idProvider := mock.NewIdentityProvider(t)
+	sp := mockp2p.NewSubscriptionProvider(t)
+
+	sv := scoring.NewSubscriptionValidator(idProvider)
+	sv.RegisterSubscriptionProvider(sp)
+
+	// mocks peer 1 not subscribed to an unknown topic.
+	peer1 := p2pfixtures.PeerIdFixture(t)
+	idProvider.On("ByPeerID", peer1).Return(unittest.IdentityFixture(), true)
+	sp.On("GetSubscribedTopics", peer1).Return([]string{"unknown-topic-1", "unknown-topic-2"})
+
+	// as peer 1 has subscribed to unknown topics, the subscription validator should return an error.
+	err := sv.CheckSubscribedToAllowedTopics(peer1)
+	require.Error(t, err)
+	require.True(t, scoring.IsInvalidSubscriptionError(err))
+}
+
+// TestSubscriptionValidator_ValidSubscription tests that when a peer has subscribed to valid
+// topics based on its Flow protocol role, the subscription validator returns no error.
+func TestSubscriptionValidator_ValidSubscriptions(t *testing.T) {
```

Comment:
Should we add one more test case where role is not a valid flow-role (observer nodes) can subscribe to public channels. Also ensure receive blocks can be subscribed to.

Reply:
Good question, the peer scoring for GossipSub is currently activated only for the _staked_ (i.e., authorized) Flow network. The cardinal principle in the _staked_ Flow network is all nodes are having a valid role (VN, AN, EN, CON, or COL). We rely on the `IdentityProvider` in the sense that all resolved identities must be valid and authorized. By a resolved identity, we mean the one with a (`true`) returned value ([here](https://github.com/onflow/flow-go/blob/8e9601184f6385de5d1005cf111d47f406e0da82/network/p2p/scoring/subscription_validator.go#L33)). We consider it a [malicious action](https://github.com/onflow/flow-go/blob/8e9601184f6385de5d1005cf111d47f406e0da82/network/p2p/scoring/subscription_validator.go#L34-L40) if a libp2p `peer.ID` belonging to the _staked_ Flow network is not resolved to an authorized identity, and we penalize it with the maximum penalty value (so that it will be muted). 

Hence, the public Flow network remains unaffected by this PR.